src/tribol/interface/tribol.cpp

```diff
@@ -583,83 +725,130 @@ int getCSRMatrix( int** I, int** J, real** vals, int csId,
                               BlockSpace row_block, 
                               BlockSpace col_block,
-                              axom::Array<integer>& row_elem_idx,
-                              axom::Array<integer>& col_elem_idx,
-                              axom::Array<mfem::DenseMatrix>& jacobians )
+                              const axom::Array<integer>* row_elem_idx,
+                              const axom::Array<integer>* col_elem_idx,
+                              const axom::Array<mfem::DenseMatrix>* jacobians )
 {
    SparseMode sparse_mode = CouplingSchemeManager::getInstance().
       getCoupling(csId)->getEnforcementOptions().lm_implicit_options.sparse_mode;
    if (sparse_mode != SparseMode::MFEM_ELEMENT_DENSE)
    {
       SLIC_WARNING("Jacobian is assembled and can be accessed by " 
          "getMfemSparseMatrix() or getCSRMatrix(). For (unassembled) element "
-         "Jacobian contributions, call setJacobianStorage(JacobianStorage::ELEMENT) "
-         "before calling update().");
+         "Jacobian contributions, call setLagrangeMultiplierOptions() with "
+         "SparseMode::MFEM_ELEMENT_DENSE before calling update().");
       return 1;
    }
    MethodData* method_data = 
       CouplingSchemeManager::getInstance().getCoupling( csId )->getMethodData();
-   row_elem_idx = method_data->getBlockJElementIds()[static_cast<int>(row_block)];
-   col_elem_idx = method_data->getBlockJElementIds()[static_cast<int>(col_block)];
-   jacobians = method_data->getBlockJ()(
+   row_elem_idx = &method_data->getBlockJElementIds()[static_cast<int>(row_block)];
+   col_elem_idx = &method_data->getBlockJElementIds()[static_cast<int>(col_block)];
+   jacobians = &method_data->getBlockJ()(
       static_cast<int>(row_block),
       static_cast<int>(col_block)
    );
    return 0;
 }
 
 //------------------------------------------------------------------------------
-void registerRealNodalField( integer meshId,
-                             const RealNodalFields field,
-                             real * fieldVariable )
+std::unique_ptr<mfem::BlockOperator> getMfemBlockJacobian( integer csId )
+{
+   CouplingScheme* coupling_scheme = CouplingSchemeManager::getInstance().
+      getCoupling(csId);
+   SparseMode sparse_mode = coupling_scheme
+      ->getEnforcementOptions().lm_implicit_options.sparse_mode;
+   if (sparse_mode != SparseMode::MFEM_ELEMENT_DENSE)
+   {
+      SLIC_ERROR_ROOT("Jacobian is assembled and can be accessed by " 
+         "getMfemSparseMatrix() or getCSRMatrix(). For (unassembled) element "
+         "Jacobian contributions, call setLagrangeMultiplierOptions() with "
+         "SparseMode::MFEM_ELEMENT_DENSE before calling update().");
+   }
+   SLIC_ERROR_ROOT_IF(
+      !coupling_scheme->hasMfemData(), 
+      "No MFEM data exists. The coupling scheme "
+      "must be registered using registerParMesh() to use this method."
+   );
+   return coupling_scheme->getMfemMatrixData()->GetMfemBlockJacobian(
```

Comment:
Consider a comment here. Unfortunately the use of MfemMatrixData and then MfemBlockJacobian, and then MethodData doesn't read super clear. I don't think we refactor the naming etc., but a comment could help.

Reply:
Comment has been added.